File: app/src/main/java/android/rr/apksapp/adapters/DashboardAdapter.java
package android.rr.apksapp.adapters;

import android.content.Context;
import android.support.v7.widget.CardView;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import android.rr.apksapp.Dashboard;
import android.rr.apksapp.R;
import android.rr.apksapp.models.AppsApkDetails;
import android.rr.apksapp.utils.GlideCircleTransform;

import java.util.ArrayList;

public class DashboardAdapter extends RecyclerView.Adapter<DashboardAdapter.ViewHolder> {
    private Context context;
    private ArrayList<AppsApkDetails> appsApkDetails;
    private boolean isLongClicked = false;
    private final String TAG = DashboardAdapter.class.getSimpleName();

    public DashboardAdapter(Context context, ArrayList<AppsApkDetails> appsApkDetails) {
        this.context = context;
        this.appsApkDetails = appsApkDetails;
    }

    public void updateAdapter(Context context, ArrayList<AppsApkDetails> appsApkDetails) {
        this.context = context;
        this.appsApkDetails = appsApkDetails;

        notifyDataSetChanged();
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new ViewHolder(LayoutInflater.from(parent.getContext()).
                inflate(R.layout.for_dashboard_grid, parent, false));
    }

    @Override
    public void onBindViewHolder(ViewHolder viewHolder, int position) {
        viewHolder.bindData(position);
    }

    @Override
    public int getItemCount() {
        return appsApkDetails.size();
    }

    class ViewHolder extends RecyclerView.ViewHolder implements
            View.OnClickListener, View.OnLongClickListener {
        CardView cardView;
        TextView appNameTV;
        ImageView appIconIV;
        int clickedPosition;

        private ViewHolder(View itemView) {
            super(itemView);

            cardView = itemView.findViewById(R.id.cardView);
            appNameTV = itemView.findViewById(R.id.appNameTV);
            appIconIV = itemView.findViewById(R.id.appIconIV);
        }

        public void bindData(int position) {
            this.clickedPosition = position;
            Glide.with(context).load("").placeholder(appsApkDetails.get(position)
                    .getDrawable()).transform(new GlideCircleTransform(context)).
                    into(appIconIV);
            appNameTV.setText(appsApkDetails.get(position).getAppName());

            if (appsApkDetails.get(position).getIsSelected()) {
                itemView.setBackgroundColor(context.getResources().
                        getColor(R.color.cardSelected));
            } else {
                itemView.setBackgroundColor(context.getResources().
                        getColor(R.color.cardUnSelected));
            }

            itemView.setOnClickListener(this);
            itemView.setOnLongClickListener(this);
        }

        @Override
        public void onClick(View view) {
            Log.d(TAG, "onClick(), clickedPosition: " + clickedPosition +
                    ", isLongClicked: " + isLongClicked);

            if (isLongClicked) {
                if (appsApkDetails.get(clickedPosition).getIsSelected()) {
                    appsApkDetails.get(clickedPosition).setIsSelected(false);
                    itemView.setBackgroundColor(context.getResources().
                            getColor(R.color.cardUnSelected));
                } else {
                    appsApkDetails.get(clickedPosition).setIsSelected(true);
                    itemView.setBackgroundColor(context.getResources().
                            getColor(R.color.cardSelected));
                }
            }

            if (0 == getSelectedItemsCount()) {
                isLongClicked = false;
                ((Dashboard) context).showOrHideBottomBar(false);
            }
        }

        @Override
        public boolean onLongClick(View view) {
            Log.d(TAG, "onLongClick(), clickedPosition: " + clickedPosition +
                    ", isLongClicked: " + isLongClicked);

            if (!isLongClicked) {
                isLongClicked = true;
                ((Dashboard) context).showOrHideBottomBar(true);

                appsApkDetails.get(clickedPosition).setIsSelected(true);
                itemView.setBackgroundColor(context.getResources().
                        getColor(R.color.cardSelected));
            }

            return true;
        }
    }

    private int getSelectedItemsCount() {
        int selectedItemsCount = 0;
        for (int i = 0; i < appsApkDetails.size(); i++) {
            if (appsApkDetails.get(i).getIsSelected()) {
                selectedItemsCount++;
            }
        }

        Log.d(TAG, "getSelectedItemsCount(), selectedItemsCount: " + selectedItemsCount);
        return selectedItemsCount;
    }

    public ArrayList<AppsApkDetails> getTheSelectedAppsList() {
        ArrayList<AppsApkDetails> arrayList = new ArrayList<>();
        arrayList.addAll(appsApkDetails);

        for (int i = 0; i < arrayList.size(); i++) {
            if (!arrayList.get(i).getIsSelected()) {
                arrayList.remove(i);
                i--;
            }
        }

        return arrayList;
    }

}

File: app/src/main/java/android/rr/apksapp/adapters/DashboardAdapter.kt
package android.rr.apksapp.adapters

import android.content.Context
import android.rr.apksapp.Dashboard
import android.rr.apksapp.R
import android.rr.apksapp.models.AppsApkDetails
import android.rr.apksapp.utils.GlideCircleTransform
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnLongClickListener
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import com.bumptech.glide.Glide
import java.util.*

class DashboardAdapter(private var context: Context?, private var appsApkDetails: ArrayList<AppsApkDetails?>?) : RecyclerView.Adapter<DashboardAdapter.ViewHolder?>() {
    private var isLongClicked = false
    private val TAG = DashboardAdapter::class.java.simpleName

    fun updateAdapter(context: Context?, appsApkDetails: ArrayList<AppsApkDetails?>?) {
        this.context = context
        this.appsApkDetails = appsApkDetails
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, p1: Int): ViewHolder {
        //TODO("Not yet implemented")
        return ViewHolder(LayoutInflater.from(parent.context).inflate(
                R.layout.for_dashboard_grid, parent, false))
    }

    override fun onBindViewHolder(viewHolder: ViewHolder, position: Int) {
        //TODO("Not yet implemented")
        viewHolder.bindData(position)
    }

    override fun getItemCount(): Int {
        return appsApkDetails?.size!!
    }

    inner class ViewHolder(itemView: View?) : RecyclerView.ViewHolder(itemView!!),
            View.OnClickListener, OnLongClickListener {
        var appNameTV: TextView? = itemView?.findViewById(R.id.appNameTV)
        var appIconIV: ImageView? = itemView?.findViewById(R.id.appIconIV)
        var clickedPosition = 0

        fun bindData(position: Int) {
            clickedPosition = position
            Glide.with(context).load("").placeholder(appsApkDetails?.get(position)
                    ?.drawable).transform(GlideCircleTransform(context)).into(appIconIV)
            appNameTV?.text = appsApkDetails?.get(position)?.appName
            if (appsApkDetails?.get(position)?.isSelected!!) {
                context?.resources?.getColor(R.color.cardSelected)?.let { itemView.setBackgroundColor(it) }
            } else {
                context?.resources?.getColor(R.color.cardUnSelected)?.let { itemView.setBackgroundColor(it) }
            }
            itemView.setOnClickListener(this)
            itemView.setOnLongClickListener(this)
        }

        override fun onClick(view: View?) {
            Log.d(TAG, "onClick(), clickedPosition: " + clickedPosition +
                    ", isLongClicked: " + isLongClicked)
            if (isLongClicked) {
                if (appsApkDetails?.get(clickedPosition)?.isSelected!!) {
                    appsApkDetails!![clickedPosition]?.isSelected =  false
                    context?.resources?.getColor(R.color.cardUnSelected)?.let { itemView.setBackgroundColor(it) }
                } else {
                    appsApkDetails?.get(clickedPosition)?.isSelected = true
                    context?.resources?.getColor(R.color.cardSelected)?.let { itemView.setBackgroundColor(it) }
                }
            }
            if (0 == getSelectedItemsCount()) {
                isLongClicked = false
                (context as Dashboard?)?.showOrHideBottomBar(false)
            }
        }

        override fun onLongClick(view: View?): Boolean {
            Log.d(TAG, "onLongClick(), clickedPosition: " + clickedPosition +
                    ", isLongClicked: " + isLongClicked)
            if (!isLongClicked) {
                isLongClicked = true
                (context as Dashboard?)?.showOrHideBottomBar(true)
                appsApkDetails?.get(clickedPosition)?.isSelected = true
                context?.resources?.getColor(R.color.cardSelected)?.let { itemView.setBackgroundColor(it) }
//                itemView.setBackgroundColor(context?.resources?.getColor(R.color.cardSelected))
            }
            return true
        }

    }

    private fun getSelectedItemsCount(): Int {
        var selectedItemsCount = 0
        for (i in appsApkDetails?.indices!!) {
            if (appsApkDetails!![i]?.isSelected!!) {
                selectedItemsCount++
            }
        }

        Log.d(TAG, "getSelectedItemsCount(), selectedItemsCount: $selectedItemsCount")
        return selectedItemsCount
    }

    fun getTheSelectedAppsList(): ArrayList<AppsApkDetails?>? {
        val arrayList = ArrayList<AppsApkDetails?>()
        appsApkDetails?.let { arrayList.addAll(it) }
//        arrayList.addAll(appsApkDetails)
        var i = 0
        while (i < arrayList.size) {
            if (!arrayList[i]?.isSelected!!) {
                arrayList.removeAt(i)
                i--
            }
            i++
        }
        return arrayList
    }

}